File: vscode-extension/test/parser.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert';
import path from 'node:path';
import { parseSessionMetadata, parseSessionExchanges } from '../src/parser';

const fixturePath = path.join(
  import.meta.dirname || __dirname,
  'fixtures/projects/-home-test-project/aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee.jsonl'
);

describe('parseSessionMetadata', () => {
  it('extracts session metadata from JSONL', () => {
    const session = parseSessionMetadata({
      sessionId: 'aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee',
      projectDir: '/home/test/project',
      jsonlPath: fixturePath,
      fileMtime: 1000,
    });
    assert.strictEqual(session.sessionId, 'aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee');
    assert.strictEqual(session.cwd, '/home/test/project');
    assert.strictEqual(session.name, 'help me fix the login bug');
    assert.strictEqual(session.version, '2.1.71');
    assert.strictEqual(session.messageCount, 4);
    assert.ok(session.firstUserMessage.includes('login bug'));
    assert.ok(session.firstAssistantMessage.includes('take a look'));
  });
});

describe('parseSessionExchanges', () => {
  it('returns all user/assistant exchanges', () => {
    const exchanges = parseSessionExchanges(fixturePath);
    assert.strictEqual(exchanges.length, 4);
    assert.strictEqual(exchanges[0].role, 'user');
    assert.ok(exchanges[0].text.includes('login bug'));
    assert.strictEqual(exchanges[1].role, 'assistant');
    assert.ok(exchanges[1].text.includes('take a look'));
  });
});

File: tests/parser_test.rs
use std::path::PathBuf;

#[path = "../src/session.rs"]
mod session;
#[path = "../src/scanner.rs"]
mod scanner;
#[path = "../src/parser.rs"]
mod parser;

#[test]
fn test_parse_session() {
    let jsonl_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/fixtures/projects/-home-test-project/aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee.jsonl");

    let discovered = scanner::DiscoveredSession {
        session_id: "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee".to_string(),
        project_dir: "/home/test/project".to_string(),
        jsonl_path: jsonl_path.to_string_lossy().to_string(),
        file_mtime: 1000,
    };

    let session = parser::parse_session(&discovered).unwrap();
    assert_eq!(session.session_id, "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee");
    assert_eq!(session.cwd, "/home/test/project");
    assert_eq!(session.message_count, 4); // 2 user + 2 assistant
    assert!(session.user_messages.contains("login bug"));
    assert!(session.user_messages.contains("middleware"));
    assert!(session.assistant_messages.contains("auth module"));
    assert_eq!(session.first_user_message, "help me fix the login bug");
    assert!(session.first_assistant_message.contains("take a look"));
    assert_eq!(session.name, "help me fix the login bug");
    assert_eq!(session.version, "2.1.71");
}

#[test]
fn test_parse_skips_non_message_types() {
    let jsonl_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/fixtures/projects/-home-test-project/aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee.jsonl");

    let discovered = scanner::DiscoveredSession {
        session_id: "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee".to_string(),
        project_dir: "/home/test/project".to_string(),
        jsonl_path: jsonl_path.to_string_lossy().to_string(),
        file_mtime: 1000,
    };

    let session = parser::parse_session(&discovered).unwrap();
    assert_eq!(session.message_count, 4);
}
